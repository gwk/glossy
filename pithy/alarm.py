# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

import signal
from typing import Callable


class Timeout(Exception):
  'Exception indictaing that an AlarmManager timed out.'


class AlarmManager:

  def __init__(self, timeout: int, msg:str='alarm timed out after {timeout} seconds', on_signal: Callable[[], None]=None) -> None:
    if not isinstance(timeout, int): raise TypeError(f'timeout must be an int; received: {timeout!r}')
    if timeout < 0: raise TypeError(f'timeout must be nonnegative; received: {timeout!r}')
    self.timeout = timeout
    self.msg = msg
    self.on_signal = on_signal
    self.timed_out = False


  def __enter__(self):
    if not self.timeout: return
    prev = signal.getsignal(signal.SIGALRM)
    if prev is not signal.SIG_DFL:
      raise Exception(f'task.run encountered previously installed signal handler: {prev}')

    def alarm_handler(signum, current_stack_frame):
      # since signal handlers carry reentrancy concerns; do not do any IO within the handler.
      self.timed_out = True
      if self.on_signal: self.on_signal() # callback must respect reentrancy limitations.

    signal.signal(signal.SIGALRM, alarm_handler) # set handler.
    signal.alarm(self.timeout) # set alarm.


  def __exit__(self, exc_type, exc_value, traceback) -> None:
    if not self.timeout: return None
    signal.alarm(0) # disable alarm.
    signal.signal(signal.SIGALRM, signal.SIG_DFL) # uninstall handler.
    if exc_type: return None # exception will be reraised.
    if self.timed_out: raise Timeout(self.msg.format(timeout=self.timeout))
    return None
