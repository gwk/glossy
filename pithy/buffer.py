# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from typing import Callable, Generic, Iterable, Iterator, List, TypeVar, Union
from .default import Raise, RaiseOr


T = TypeVar('T')


class Buffer(Iterator[T]):
  '''
  Iterable object that buffers an iterable.
  Call push() to push an item into the buffer;
  this will be returned on the subsequent call to next().
  '''


  def __init__(self, iterable: Iterable[T]) -> None:
    self.iterator = iter(iterable)
    self.buffer: List[T] = []


  def __repr__(self) -> str:
    return 'IterBuffer({!r}, buffer={!r})'.format(self.iterator, self.buffer)


  def __iter__(self) -> Iterator[T]: return self


  def __next__(self) -> T:
    try: return self.buffer.pop()
    except IndexError: pass
    return next(self.iterator)


  @property
  def is_live(self) -> bool:
    try: self.peek()
    except StopIteration: return False
    else: return True


  def push(self, item: T) -> None:
    self.buffer.append(item)


  def peek(self, default: RaiseOr[T]=Raise._) -> T:
    try: return self.buffer[-1]
    except IndexError: pass
    try: el = next(self.iterator)
    except StopIteration:
      if isinstance(default, Raise): raise
      else: return default
    self.buffer.append(el)
    return el


  def take_while(self, pred: Callable[[T], bool]) -> Iterator[T]:
    for el in self:
      if pred(el):
        yield el
      else:
        self.buffer.append(el)
        break


  def drop_while(self, pred: Callable[[T], bool]) -> None:
    for el in self:
      if not pred(el):
        self.buffer.append(el)
        break


  def peek_while(self, pred: Callable[[T], bool]) -> List[T]:
    els = list(self.take_while(pred))
    self.buffer.extend(reversed(els))
    return els


  def take(self, count: int, short=False, default: RaiseOr[T]=Raise._) -> List[T]:
    els = []
    for _ in range(count):
      try: els.append(next(self))
      except StopIteration:
        if short: break
        if isinstance(default, Raise): raise
        els.append(default)
    return els


  def peeks(self, count: int, short=False, default: RaiseOr[T]=Raise._) -> List[T]:
    if 0 < count <= len(self.buffer):
      return list(reversed(self.buffer[-count:]))
    els = []
    for _ in range(count):
      try: els.append(next(self))
      except StopIteration:
        if short: break
        if isinstance(default, Raise): raise
        els.append(default)
    self.buffer.extend(reversed(els))
    return els


  def expect(self, pred: Callable[[T], bool]) -> T:
    el = next(self)
    if pred(el): return el
    raise ValueError(el)
