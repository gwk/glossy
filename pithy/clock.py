# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from sys import stderr
from time import time as now


class Clock:
  def __init__(self, name:str):
    self.name = name
    self.times = [(now(), 'start')]

  def split(self, label:str):
    t = now()
    print(f'clock: {self.name}: {label}: {t - self.times[-1][0]:.04f}', file=stderr)
    self.times.append((t, label))


  def total(self):
    t = now()
    print(f'clock: {self.name}: TOTAL: {t - self.times[0][0]:.04f}', file=stderr)
