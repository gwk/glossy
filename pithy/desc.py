# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

import re
from sys import stderr, stdout
from typing import Any, Iterable, Iterator, List, Mapping, Optional, Set, TextIO, Tuple

from .tree import known_leaf_types


def writeD(file:TextIO, *labels_and_obj:Any, depth:int=0) -> None:
  obj = labels_and_obj[-1]
  labels = labels_and_obj[:-1]
  if labels: print(*labels, end=': ', file=file)
  for line in gen_desc(obj, depth=depth):
    print(line, file=file)


def errD(*labels_and_obj:Any, depth:int=0) -> None:
  writeD(stderr, *labels_and_obj, depth=depth)


def outD(*labels_and_obj:Any, depth:int=0) -> None:
  writeD(stdout, *labels_and_obj, depth=depth)


def gen_desc(obj:Any, depth:int=0) -> Iterator[str]:
  width = 128
  buffer:List[str] = []
  buffer_depth = depth

  def needs_multiline() -> bool:
    return False

  def flush(multiline:bool=False) -> Iterator[str]:
    ind = '  ' * buffer_depth
    if multiline or needs_multiline():
      last = len(buffer) - 1
      for i, el in enumerate(buffer):
        comma = ',' if i < last else ''
        yield f'{ind}{el}{comma}'
    else: # Inline.
      yield ind + ', '.join(buffer)
    buffer.clear()

  for d, s in gen_obj_desc(obj, depth=depth, visited_ids=[]):
    if d < 0: # Closer.
      assert buffer
      buffer[-1] += s if (buffer[-1][-1] in '[]{}()') else ' '+s
    elif d < buffer_depth: # Pop level out.
      yield from flush()
      buffer_depth = d
      buffer.append(s)
    elif d > buffer_depth: # Push level in.
      yield from flush(multiline=True)
      buffer_depth = d
      buffer.append(s)
    else: # Same depth.
      buffer.append(s)

  yield from flush()


def gen_obj_desc(obj:Any, depth:int, visited_ids:List[int]) -> Iterator[Tuple[int,str]]:
  if isinstance(obj, known_leaf_types):
    yield (depth, repr(obj))
    return

  i = id(obj)
  if i in visited_ids:
    yield (depth, f'^0x{i:x}:{type(obj).__name__}')
    return

  # Most objects in a tree are leaves; we minimize tests for the leaf case by nesting the mapping test inside the iter test.
  # This has the side effect of ignoring non-iterable classes that have an irrelevant items() function.

  # Explicit test because iter() will return for objects without __iter__ but with __getitem__;
  # this includes typing._SpecialForm, which fails when we attempt to iterate.
  if hasattr(obj, '__iter__'):
    try: it = iter(obj)
    except TypeError: pass
    else:
      visited_ids.append(i)
      # Attempt to distinguish between mapping and sequence types.
      try: items = iter(obj.items()) # Wrapping `iter` guards against badly formed items() functions.
      except (AttributeError, TypeError): # Treat as iterable.
        yield from gen_iter_desc(obj, iter(obj), depth, visited_ids)
      else: # Treat as a mapping.
        yield from gen_mapping_desc(obj, items, depth, visited_ids)
      visited_ids.pop()
      return

  yield (depth, repr(obj))


def gen_mapping_desc(obj:Mapping, items:Iterator[Tuple[Any,Any]], depth:int, visited_ids:List[int]) -> Iterator[Tuple[int,str]]:
  is_dict = isinstance(obj, dict)
  head = '{' if is_dict else (type(obj).__qualname__ + '({')
  yield (depth, head)
  for pair in items:
    try: k, v = pair
    except (TypeError, ValueError):
      yield (depth+1, f'! {pair!r}')
      continue
    vg = gen_obj_desc(v, depth+1, visited_ids)
    v1d, v1s = next(vg)
    ks = f'{k!r}: {v1s}'
    yield (depth+1, ks)
    yield from vg
  yield (-1, '}' if is_dict else '})')


def gen_iter_desc(obj:Any, it:Iterator, depth:int, visited_ids:List[int]) ->  Iterator[Tuple[int,str]]:
  if isinstance(obj, list):
    head = '['
    close = ']'
  elif isinstance(obj, tuple):
    head = '('
    close = ',)' if len(obj) == 1 else ')'
  else:
    head = type(obj).__qualname__ + '(['
    close = '])'
  yield (depth, head)
  for el in it: yield from gen_obj_desc(el, depth+1, visited_ids)
  yield (-1, close)


def repr_clean(obj:Any) -> str:
  '''
  Attempt to ensure that an object gets printed with a decent repr.
  Some third-party libraries print unconventional reprs that are confusing inside of collection descriptions.
  '''
  r = repr(obj)
  if type(obj) in _literal_repr_types or _decent_repr_re.fullmatch(r): return r
  return f'{type(obj).__name__}({r})'

_literal_repr_types = { bytes,str,list,dict,set,tuple }
_decent_repr_re = re.compile(r'[a-zA-Z][.\w]*\(.*\)')


def repr_lim(obj:Any, limit=64) -> str:
  r = repr(obj)
  if len(r) <= limit: return r
  return r[:limit-1] + 'â€¦'
