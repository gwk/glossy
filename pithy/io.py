# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from pprint import pprint
from sys import argv, stdin, stdout, stderr
from string import Template
from typing import Any, Iterable, Iterator, TextIO, TypeVar, Union


_T = TypeVar('_T')

# basic printing.

def writeZ(file: TextIO, *items: Any, sep='', end='', flush=False) -> None:
  "Write `items` to file; default sep='', end=''."
  print(*items, sep=sep, end=end, file=file, flush=flush)

def writeS(file: TextIO, *items: Any, sep='', flush=False) -> None:
  "Write `items` to file; sep='', end=' '."
  print(*items, sep=sep, end=' ', file=file, flush=flush)

def writeSZ(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep=' ', end=''."
  print(*items, sep=' ', end='', file=file, flush=flush)

def writeSS(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep=' ', end=''."
  print(*items, sep=' ', end=' ', file=file, flush=flush)

def writeL(file: TextIO, *items: Any, sep='', flush=False) -> None:
  "Write `items` to file; sep='', end='\\n'."
  print(*items, sep=sep, end='\n', file=file, flush=flush)

def writeN(file: TextIO, *items: Any, sep='', flush=False) -> None:
  "Write `items` to file; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=sep, end=end, file=file, flush=flush)

def writeSL(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep=' ', end='\\n'."
  print(*items, sep=' ', end='\n', file=file, flush=flush)

def writeSN(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=' ', end=end, file=file, flush=flush)

def writeLL(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep='\\n', end='\\n'."
  print(*items, sep='\n', end='\n', file=file, flush=flush)

def writeLSSL(file: TextIO, *items: Any, flush=False) -> None:
  "Write `items` to file; sep='\\n  ', end='\\n'."
  print(*items, sep='\n  ', end='\n', file=file, flush=flush)


# templated format printing.

def writeTF(file: TextIO, template_fmt, *items: Any, flush=False, **keyed_items: Any) -> None:
  """
  Expand the format string with keyed_items, then format the string; end=''.
  Useful for constructing dynamic format strings.
  """
  fmt = Template(template_fmt).substitute(**keyed_items)
  print(fmt.format(*items, **keyed_items, end='', file=file, flush=flush))


def writeTFL(file: TextIO, template_fmt, *items: Any, flush=False, **keyed_items: Any) -> None:
  """
  Expand the format string template with keyed_items, then format the string; end='\\n'
  Useful for constructing dynamic format strings.
  """
  fmt = Template(template_fmt).substitute(**keyed_items)
  print(fmt.format(*items, **keyed_items, file=file, flush=flush))


def writeP(file: TextIO, *items: Any, label=None, indent=2, **opts: Any) -> None:
  'Pretty print to file.'
  if label is not None:
    file.write(label)
    file.write (': ')
  for item in items:
    pprint(item, stream=file, indent=indent, **opts)


# std out.

def outZ(*items: Any, sep='', end='', flush=False) -> None:
  "Write `items` to std out; sep='', end=''."
  print(*items, sep=sep, end=end, flush=flush)

def outS(*items: Any, sep='', flush=False) -> None:
  "Write `items` to std out; sep='', end=' '."
  print(*items, sep=sep, end=' ', flush=flush)

def outSZ(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep=' ', end=''."
  print(*items, sep=' ', end='', flush=flush)

def outSS(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep=' ', end=' '."
  print(*items, end=' ', flush=flush)

def outL(*items: Any, sep='', flush=False) -> None:
  "Write `items` to std out; sep='', end='\\n'."
  print(*items, sep=sep, flush=flush)

def outN(*items: Any, sep='', flush=False) -> None:
  "Write `items` to std out; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=sep, end=end, flush=flush)

def outSL(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep=' ', end='\\n'."
  print(*items, flush=flush)

def outSN(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=' ', end=end, flush=flush)

def outLL(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep='\\n', end='\\n'."
  print(*items, sep='\n', flush=flush)

def outLSSL(*items: Any, flush=False) -> None:
  "Write `items` to std out; sep='\\n  ', end='\\n'."
  print(*items, sep='\n  ', flush=flush)

def outP(*items: Any, label=None, flush=False, **opts: Any) -> None:
  'Pretty print to std out.'
  writeP(stdout, *items, label=label, **opts)


# std err.

def errZ(*items: Any, sep='', end='', flush=False) -> None:
  "Write items to std err; default sep='', end=''."
  print(*items, sep=sep, end=end, file=stderr, flush=flush)

def errS(*items: Any, sep='', flush=False) -> None:
  "Write items to std err; sep='', end=' '."
  print(*items, sep=sep, end=' ', file=stderr, flush=flush)

def errSZ(*items: Any, flush=False) -> None:
  "Write items to std err; sep=' ', end=''."
  print(*items, sep=' ', end='', file=stderr, flush=flush)

def errSS(*items: Any, flush=False) -> None:
  "Write items to std err; sep=' ', end=''."
  print(*items, sep=' ', end=' ', file=stderr, flush=flush)

def errL(*items: Any, sep='', flush=False) -> None:
  "Write items to std err; sep='', end='\\n'."
  print(*items, sep=sep, end='\n', file=stderr, flush=flush)

def errN(*items: Any, sep='', flush=False) -> None:
  "Write `items` to std err; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=sep, end=end, file=stderr, flush=flush)

def errSL(*items: Any, flush=False) -> None:
  "Write items to std err; sep=' ', end='\\n'."
  print(*items, sep=' ', end='\n', file=stderr, flush=flush)

def errSN(*items: Any, flush=False) -> None:
  "Write `items` to std err; sep='', end=('' if items and items[-1].endswith('\\n') else '\\n')."
  end = ('' if items and items[-1].endswith('\n') else '\n')
  print(*items, sep=' ', end=end, file=stderr, flush=flush)

def errLL(*items: Any, flush=False) -> None:
  "Write items to std err; sep='\\n', end='\\n'."
  print(*items, sep='\n', end='\n', file=stderr, flush=flush)

def errLSSL(*items: Any, flush=False) -> None:
  "Write items to std err; sep='\\n  ', end='\\n'."
  print(*items, sep='\n  ', end='\n', file=stderr, flush=flush)

def errP(*items: Any, label=None, **opts) -> None:
  'Pretty print to std err.'
  writeP(stderr, *items, label=label, **opts)


def err_progress(iterable: Iterable[_T], label='progress', suffix='', final_suffix='', frequency:Union[float, int]=0.1, limit=0) -> Iterator[_T]:
  '''
  For interactive terminals, return a generator that yields the elements of `iterable`
  and displays a progress indicator on std err.
  '''
  assert frequency >= 0
  if not frequency or not stderr.isatty():
    return iter(iterable)

  ERASE_LINE = '\x1b[2K'

  if label is None:
    label = str(iterable)
  pre = f'{ERASE_LINE}\r◊ {label}: '
  post = (suffix and ' ' + suffix) + '…'
  final = f' {final_suffix}.' if final_suffix else '.'

  if isinstance(frequency, float):
    from time import time
    def err_progress_gen() -> Iterator[_T]:
      prev_t = time()
      step = 1
      next_i = step
      i = -1
      for i, el in enumerate(iterable):
        if limit and i == limit:
          i -= 1
          break
        if i == next_i:
          print(f'{pre}{i:,}{post}', end='', file=stderr, flush=True)
          t = time()
          d = t - prev_t
          step = max(1, int(step * frequency / d))
          prev_t = t
          next_i = i + step
        yield el
      print(f'{pre}{i+1:,}{final}', file=stderr)


  else:
    def err_progress_gen() -> Iterator[_T]:
      for i, el in enumerate(iterable):
        if limit and i == limit:
          i -= 1
          break
        if i % frequency == 0:
          print(pre + str(i) + post, end='', file=stderr, flush=True)
        yield el
      print(pre + str(i) + final, file=stderr)

  return err_progress_gen()


# convenience read/write.


def read_from_path(path: str, default: str=None) -> str:
  'Read all text from file at `path`.'
  try:
    with open(path) as f:
      return f.read()
  except (FileNotFoundError, IsADirectoryError):
    if default is None: raise
    return default

def read_line_from_path(path: str, line0=0, keep_end=False, default: str=None) -> str:
  'Read a single line of text from file at `path`.'
  try:
    with open(path) as f:
      for i, line in enumerate(f):
        if i == line0:
          return line if keep_end else line.rstrip('\n')
      if default is None: raise IndexError(line0)
      return default
  except (FileNotFoundError, IsADirectoryError, PermissionError, UnicodeDecodeError):
    if default is None: raise
    return default

def write_to_path(path: str, string) -> None:
  'Writes `string` to file at `path`.'
  with open(path, 'w') as f:
    f.write(string)


# misc.

def clip_newlines(iterable: Iterable[str]) -> Iterable[str]:
  for line in iterable:
    yield line.rstrip('\n')


def confirm(question:str) -> bool:
  from .term import change_mode, CBREAK
  print(f'{question}? press "y" to confirm: ', end='', flush=True)
  with change_mode(CBREAK):
    response = stdin.read(1)
    print(response)
    return (response == 'y')


def assert_eq(a: Any, b: Any):
  if a != b:
    raise AssertionError(f'not equal:\n  {a!r}\n  {b!r}')


def tee_to_err(iterable:Iterable[_T], label:str = 'tee_to_err') -> Iterator[_T]:
  for el in iterable:
    errL(label, ': ', repr(el))
    yield el
