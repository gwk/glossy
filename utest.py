'''
utest is a tiny unit testing library.
'''


import atexit
from sys import stderr


__all__ = ['utest', 'utest_exc', 'utest_val']


test_count = 0
failure_count = 0


def utest(exp, fn, *args, **kwargs):
  '''
  Invoke `fn` with `args` and `kwargs`,
  and log a test failure if an exception is raised or the returned value does not equal `exp`.
  '''
  global test_count
  test_count += 1
  try:
    ret = fn(*args, **kwargs)
    exc = None
  except BaseException as e:
    ret = None
    exc = e
  else:
    if exp == ret: return
  log_failure('value', exp=exp, ret=ret, exc=exc, name=fn.__qualname__, args=args, kwargs=kwargs)


def utest_exc(exp_exc, fn, *args, **kwargs):
  '''
  Invoke `fn` with `args` and `kwargs`,
  and log a test failure if an exception is not raised or the raised exception type and args not match `exp_exc`.
  '''
  global test_count
  test_count += 1
  try:
    ret = fn(*args, **kwargs)
    exc = None
  except BaseException as e:
    if exceptions_eq(exp_exc, e): return
    ret = None
    exc = e
  log_failure('exception', exp=exp_exc, ret=ret, exc=exc, name=fn.__qualname__, args=args, kwargs=kwargs)


def utest_val(exp_val, act_val, name):
  '''
  Log a test failure if `exp_val` does not equal `act_val`.
  '''
  global test_count
  test_count += 1
  if exp_val == act_val:
    return
  log_failure('value', exp=exp_val, ret=act_val, exc=None, name=name, args=(), kwargs={})


def log_failure(exp_prefix, exp, ret, exc, name, args, kwargs):
  global failure_count
  failure_count += 1
  msg_lines = ['utest failure: ' + name]
  def msg(fmt, *items): msg_lines.append(('  ' + fmt).format(*items))
  for i, el in enumerate(args):
    msg('arg {}={!r}', i, el)
  for name, val, in sorted(kwargs.items()):
    msg('arg {}={!r}', name, val)
  msg('expected {}: {!r}', exp_prefix, exp)
  if exc is None: # unexpected value.
    msg('returned value: {!r}', ret)
  else: # unexpected exception.
    msg('raised exception: {!r}', exc)
  print(*msg_lines, sep='\n', end='\n\n', file=stderr)


def exceptions_eq(a, b):
  'Compare two exceptions; since Python exceptions do not implement value equality; we do our best here.'
  if type(a) != type(b): return False
  return a.args == b.args


@atexit.register
def report():
  'At process exit, if any test failures occured, print a summary message and cause process to exit with status code 1.'
  from os import _exit
  if failure_count > 0:
    print('\nutest ran: {}; failed: {}'.format(test_count, failure_count), file=stderr)
    _exit(1) # raising SystemExit has no effect in an atexit handler as of 3.5.2.

