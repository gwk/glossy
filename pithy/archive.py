# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from functools import total_ordering
from io import BytesIO, TextIOBase, TextIOWrapper
from typing import Any, BinaryIO, Callable, Iterable, List, Tuple, Union
from tarfile import TarFile
from zipfile import ZipFile, BadZipFile
from .fs import path_ext
from .util import lazy_property
from .io import *


FileOrPath = Union[BinaryIO, str] # TODO: support other pathlike objects.
Opener = Callable[[], BinaryIO]


class Archive:

  def __init__(self, file_or_path:FileOrPath, ext=None) -> None:
    self.name: str = getattr(file_or_path, 'name', str(file_or_path))
    if ext is None:
      ext = path_ext(self.name)
    self.ext = ext
    try: self._handler = _ext_handlers[ext](file_or_path)
    except KeyError: raise ValueError(f'Archive does not support extension: {ext!r}')


  def __repr__(self) -> str:
    return f'{self.__class__.__name__}(name={self.name!r})'


  def __iter__(self) -> Iterable['ArchiveFile']:
    return self._handler.files(self)



@total_ordering
class ArchiveMember:

  def __init__(self, archive: Archive, name: str) -> None:
    self.archive = archive
    self.name = name

  def __repr__(self) -> str:
    return f'{self.__class__.__name__}({self.archive}, {self.name!r})'

  def __lt__(self, other: Any) -> bool:
    if isinstance(other, ArchiveMember):
      return self.name < other.name
    return NotImplemented



class ArchiveFile(ArchiveMember):

  def __init__(self, archive:Archive, name:str, opener:Opener) -> None:
    super().__init__(archive=archive, name=name)
    self._opener = opener

  @lazy_property
  def _file(self): return self._opener()

  def __iter__(self): return iter(self._file)

  def __getattr__(self, name):
    # note: dunder attribute access is never forwarded to this method.
    if name.startswith('_'): raise AttributeError(name)
    return getattr(self._file, name)

  def seekable(self):
    return getattr(self._file, 'seekable', False)

  def text(self, encoding='UTF-8', errors=None, newline=None) -> TextIOWrapper:
    'Returns a TextIOWrapper around the underlying binary buffer.'
    return TextIOWrapper(self._file, encoding=encoding, errors=errors, newline=newline)



class ArchiveDir(ArchiveMember): pass


class _Handler:
  def __init__(self, file_or_path:FileOrPath) -> None: raise NotImplementedError()


class _TarHandler(_Handler):
  def __init__(self, file_or_path:FileOrPath) -> None:
    self.tar = TarFile.open(mode='r|*', fileobj=file_or_path)

  def files(self, archive:Archive) -> Iterable[Tuple[str, Opener]]:
    for info in self.tar:
      if info.isdir(): continue
      opener = lambda: self.tar.extractfile(info)
      #^ note: crucial to pass info here and not name, or else gz archives get very slow.
      yield ArchiveFile(archive=archive, name=info.name, opener=opener)


class _ZipHandler(_Handler):
  def __init__(self, file_or_path:FileOrPath) -> None:
    try: self.zip = ZipFile(file_or_path)
    except BadZipFile as e:
      if isinstance(file_or_path, TextIOBase):
        raise TypeError(f'Archive requires a path or binary stream but received text stream: {file_or_path}') from e
      raise

  def files(self, archive:Archive) -> Iterable[Tuple[str, Opener]]:
    for name in self.zip.namelist():
      if name.endswith('/'): continue # directory.
      opener = lambda: self.zip.open(name)
      yield ArchiveFile(archive=archive, name=name, opener=opener)


_ext_handlers = {
  '.bz2' : _TarHandler,
  '.gz'  : _TarHandler,
  '.tar' : _TarHandler,
  '.xz', : _TarHandler,
  '.zip' : _ZipHandler,
}
