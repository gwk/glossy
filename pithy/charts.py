from .svg import *
from io import StringIO
from typing import *

class XYChart:
  def __init__(self, *series:Tuple[str, Iterable[Tuple]],
   title:str=None,
   x_label:str=None, y_label:str=None,
   w:int=None, h:int=None,
   plot_w:int=None, plot_h:int=None,
   x_min:int=None, x_max:int=None, y_min:int=None, y_max:int=None,
   title_height:int=None,
   label_width:int=None, label_height:int=None,
   legend_width:int=None, # TODO: legend_height would imply that the legend is at bottom.
   ) -> None:
    self.series = {k : list(v) for k, v in series}

    #take smallest_x and largest_x value from data if x_min or y_min is undefined
    min_x:float = 0
    max_x:float = 1
    min_y:float = 0
    max_y:float = 1
    for k,s in self.series.items():
      for x,y in s:
        min_x = min(x, min_x)
        max_x = max(x, max_x)
        min_y = min(y, min_y)
        max_y = max(y, max_y)
    if(x_min == None): x_min = min_x
    if(y_min == None): y_min = min_y
    if(x_max == None): x_max = max_x
    if(y_max == None): y_max = max_y
    self.x_min = x_min
    self.x_max = x_max
    self.y_min = y_min
    self.y_max = y_max
    x_range:float = x_max - x_min
    y_range:float = y_max - y_min
    self.svg_width = w
    self.svg_height = h
    self.label_height = label_height
    self.label_width = label_width
    self.legend_width = legend_width
    self.title_height = title_height
    self.title = title
    self.x_label = x_label
    self.y_label = y_label

    x_scale = ((self.svg_width - 10) - (self.legend_width + self.label_width + 10))/x_range
    y_scale = ((self.svg_height - self.title_height)-(self.label_height+10))/y_range
    scaled_min_x=0
    scaled_min_y=0
    for k,s in self.series.items():
      for x,y in s:
        scaled_min_x = min(x*x_scale, scaled_min_x)
        scaled_min_y = min(y*y_scale, scaled_min_y)
    for k,s in self.series.items():
      temp_list = []
      for x,y in s:
        temp_list.append(((x*x_scale)-scaled_min_x, (y*y_scale)-scaled_min_y))
      self.series[k] = temp_list



  def render(self, file=None) -> str:
    colors:List = ['#EF7F6D', '#F78398', '#E893C4', '#C2AAE5', '#89C2F3', '#49D5E9', '#38E2CA', '#70E9A0', '#AFEB77', '#EFE55D']
    if file is None:
      file = StringIO()
    chart_x_min:int = self.legend_width + self.label_width + 10
    chart_x_max:int = self.svg_width - 10
    chart_y_min:int = self.label_height+10
    chart_y_max:int = self.svg_height - self.title_height
    with SvgWriter(file, w=self.svg_width, h=self.svg_height) as svg:
      y_range = chart_y_max - chart_y_min
      #draw graph
      svg.rect(pos=(chart_x_min, chart_y_min), size=(chart_x_max-chart_x_min, chart_y_max - chart_y_min), stroke=None, fill='#fafafa')
      transform_x =  700
      color_picker = 0
      with svg.g(transform=f'translate({chart_x_min} {chart_y_max}) scale(1 -1) '):
        for i in range(1,5):
            svg.line((0, (y_range/5)*i), (chart_x_max-70, (y_range/5)*i), stroke='#BBCCDD' )
            with svg.g(transform=' scale(1 -1) '):
              svg.text(text=((self.y_max-self.y_min)/5*i), x = -40, y = -((y_range/5)*i) + 6, style="color:black") #max y
        for k,s in self.series.items(): #each series
          for p in s:
            print(p)
          svg.polyline((p for p in s), style=f'fill:none;stroke:{colors[color_picker]};stroke-width:1')
          color_picker = (1+color_picker) % 10
          print(color_picker)

      #text
      svg.text(text=self.title, x=self.svg_width/1.9, y = self.label_height*1.4, style=f"color:black;font-size:{self.title_height}") #title
      svg.text(text=self.y_label, x=0, y = self.svg_height/2, style="color:black") #x axis
      svg.text(text=self.x_label, x=self.svg_width/2, y = self.svg_height, style="color:black") #y axis
      svg.text(text=str(self.x_min), x = chart_x_min, y = (chart_y_max+15), style="color:black") #min x
      svg.text(text=str(self.y_min), x = chart_x_min-10, y = (chart_y_max), style="color:black") #min y
      svg.text(text=str(self.x_max), x = chart_x_max-20, y = (chart_y_max+15), style="color:black") #max x
      svg.text(text=str(self.y_max), x = chart_x_min-25, y = (chart_y_min+10), style="color:black") #max y
      #horizontal lines




      pass
