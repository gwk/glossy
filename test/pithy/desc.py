#!/usr/bin/env python3

from utest import *
from pithy.desc import *
from pithy.io import outL

l_empty = [] # Noncyclical; referenced multiple times.

objs = [
  0,
  '',
  'a',
  b'b',
  bytearray(),

  [],
  [0],
  [0, 1],
  [0, [1, 2]],
  [l_empty, l_empty, [l_empty]], # Should not backref.

  tuple(),
  (0,),
  (0, 0),
  (0, (1, 2)),

  {},
  {0:0, 1:1},
  {0:0, 1:{2:3, 4:5}},

  set(),
  {0},
  {0, 1},

  frozenset(),
]

for obj in objs:
  outL()
  outD(repr(obj), obj)
