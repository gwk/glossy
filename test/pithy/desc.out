
0: 0

'': ''

'a': 'a'

b'b': b'b'

bytearray(b''): bytearray(b'')

[]: []

[0]: [ 0 ]

[[0], [0]]: [
  [ 0 ]
  [ 0 ]]

[0, 1]: [ 0 1 ]

[0, [1, 2]]: [
  0
  [ 1 2 ]]

[[0, 1], 2]: [
  [ 0 1 ]
  2 ]

[[], [], [[]]]: [
  []
  []
  [
    []]]

(): ()

(0,): ( 0 ,)

(0, 0): ( 0 0 )

(0, (1, 2)): (
  0
  ( 1 2 ))

{}: {}

{0: 0, 1: 1}: { 0:0 1:1 }

{0: 0, 1: {2: 3, 4: 5}}: {
  0:0
  1:{ 2:3 4:5 }}

set(): set()

{0}: { 0 }

{0, 1}: { 0 1 }

frozenset(): frozenset([])

[DC(x=0, y=1), DC(x=[2, 3], y=4)]: [
  DC( x=0 y=1 )
  DC(
    x=[ 2 3 ]
    y=4 )]

Exact mode: {
  'a':[
    1,
    [ 2, 3 ]],
  4:5 }
